test/ssltest.c

```diff
@@ -1829,24 +1876,53 @@ int main(int argc, char *argv[])
     if (sn_server1 != NULL || sn_server2 != NULL)
         SSL_CTX_set_tlsext_servername_callback(s_ctx, servername_cb);
 
-    c_ssl = SSL_new(c_ctx);
-    s_ssl = SSL_new(s_ctx);
+    BIO_printf(bio_stdout, "Doing handshakes=%d bytes=%ld\n", number, bytes);
+    for (i = 0; i < number; i++) {
+        if (i == 0 || tls1_reuse) {
+            SSL_SESSION *sess = NULL;
 
-    if (sn_client)
-        SSL_set_tlsext_host_name(c_ssl, sn_client);
+            /* Server downgrade */
+            if (tls1_reuse && i != 0) {
+                SSL_CTX_set_options(s_ctx, SSL_OP_NO_TLSv1_1);
+                SSL_CTX_set_options(s_ctx, SSL_OP_NO_TLSv1_2);
 
-    if (!set_protocol_version(server_min_proto, s_ssl, SSL_CTRL_SET_MIN_PROTO_VERSION))
-        goto end;
-    if (!set_protocol_version(server_max_proto, s_ssl, SSL_CTRL_SET_MAX_PROTO_VERSION))
-        goto end;
-    if (!set_protocol_version(client_min_proto, c_ssl, SSL_CTRL_SET_MIN_PROTO_VERSION))
-        goto end;
-    if (!set_protocol_version(client_max_proto, c_ssl, SSL_CTRL_SET_MAX_PROTO_VERSION))
-        goto end;
+                sess = dup_session(SSL_get_session(c_ssl));
```

Comment:
I want to reset method and every internal field in session too, but will give a try to `SSL_SESSION_up_ref()`. It is very likely that it will work just fine too.